tests/testswitch.nim

```diff
@@ -209,6 +210,76 @@ suite "Switch":
 
     waitFor(testSwitch())
 
+  test "e2e should prevent more than maximum conns (listen)":
+    proc testSwitch() {.async, gcsafe.} =
+      var awaiters: seq[Future[void]]
+
+      let lestener = newStandardSwitch(maxConns = 2)
+      awaiters.add(await lestener.start())
+
+      let switches = @[
+        newStandardSwitch(),
+        newStandardSwitch(),
+        newStandardSwitch()]
+
+      var triggered: bool
+      for i, s in switches:
+        try:
+          await s.connect(
+            lestener.peerInfo.peerId,
+            lestener.peerInfo.addrs).wait(100.millis)
+
+          check i < 2 # shouldn't dial more than twice
+        except AsyncTimeoutError:
+          check i == 2
+          triggered = true
+
+      check triggered
+
+      await allFuturesThrowing(
+        switches
+        .mapIt( it.stop() )
+        .concat( @[lestener.stop()] )
+      )
+      await allFuturesThrowing(awaiters)
+
+    waitFor(testSwitch())
+
+  test "e2e should prevent more than max conns (dialer)":
```

Comment:
since cancelling is prominent in this PR, cancelling a blocked dial would really need a test here to ensure that the connmanager is properly cleaned up etc

Reply:
True, cancelation tests need to be added all the way, my worry is bloating this PR even more, it's already quite a large change as it is.

Reply:
That said, maybe the tests are crucial, I'll try to add tose.

Reply:
yes, I would say they are, this branch significantly depends on cancelling things